### PR TITLE
Scottx611x/read input from url

### LINKED ISSUE #3: High dependance on file naming
See here: https://github.com/scottx611x/refinery-higlass-docker/blob/master/on_startup.py#L48

Files probably don't always need to have `multires` in their filename

### LINKED ISSUE #6: Use the newest higlass
- This project is pretty out of date, theres been a lot of development on [higlass](https://github.com/hms-dbmi/higlass) since I've worked on this initial implementation. It would be good to spend some time and harden this tool.

- Have some sort of automated build process that will create and tag new `refinery-higlass-docker` images when Higlass itself is updated.

### PR DESCRIPTION
**Pull request by scottx611x**
@mccalluc This allows reading from `INPUT_JSON_URL`, not bringing up the nginx until tilesets have been ingested, and some general tidying.

Tests pass, but there are still only a subset of tracks selectable in the ui which I'll talk to pete about.

Just wanted some feedback before this gets too far along.

Fix #3 
Fix #6 

Dependent on: https://github.com/refinery-platform/refinery-platform/pull/2677

TODO:
- [x] ~~~Be able to handle all inputs that higlass can currently ingest~~~
  - ~~~See: [here](https://hms-dbmi.github.io/higlass-docs/data_preparation.html#bedpe-like-files) & [here](https://github.com/hms-dbmi/higlass-server/blob/master/tilesets/generate_tiles.py#L887:37)~~~
**EDIT: It wasn't feasible to infer HiGlass filetypes/datatypes in an exhaustive manner. So we are only going to accept **multires cooler** files just as the 4DN team is currently doing.
- [ ] Update higlass vis tool annotation to only list COOLER RE: https://github.com/refinery-platform/refinery-platform/issues/2358

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
+FROM gehlenborglab/higlass:v0.2.61
 
 COPY on_startup.py /home/higlass/projects/higlass-server
-COPY input.json $HIGLASS_SERVER_BASE_DIR
 
-# Display our waiting page until higlass ingests all tilesets
-COPY index.html /home/higlass/projects/higlass-website/index.html
+
+# Swap the "app" html with the main html to always provide the /app/ view
+RUN cp /home/higlass/projects/higlass-website/app/index.html /home/higlass/projects/higlass-website/index.html
```
Comment (mccalluc) [inline Dockerfile]: Would you just want to do a batch copy and maybe get the static assets, etc?

File: on_startup.py [hunk 2]
```diff
@@ -1,37 +1,69 @@
...
 
-logger = logging.getLogger(__name__)
+HIGLASS_DATA_TYPE = "higlass_data_type"
+HIGLASS_FILE_TYPE = "higlass_file_type"
+
+TILE_SETS = {}
+
+
+def get_refinery_input_as_json():
+    return requests.get(os.environ["INPUT_JSON_URL"]).json()
```
Comment (mccalluc) [inline on_startup.py]: Generally, we don't want to litter the code with error checking, but this might be an ok place, since we know there are other modes, and better feedback might be valuable... maybe getattr and then make sure it's not None?

**Review by mccalluc — COMMENTED** [review]
Will come back to this, but first impression is good.

File: on_startup.py [hunk 3]
```diff
@@ -1,37 +1,69 @@
...
 from django.core.management import call_command
 
-logger = logging.getLogger(__name__)
+HIGLASS_DATA_TYPE = "higlass_data_type"
+HIGLASS_FILE_TYPE = "higlass_file_type"
+
+TILE_SETS = {}
+
+
+def get_refinery_input_as_json():
```
Comment (mccalluc) [inline on_startup.py]: Maybe get rid of `_as_json`: Just reading the method name, I might think this returns a string containing json.

File: on_startup.py [hunk 4]
```diff
@@ -1,37 +1,69 @@
...
+    Download remote files specified by urls in the data provided by a GET to
+    the provided INPUT_JSON_URL
     :param data_dir: <String> Path to directory to populate with data
     """
-    with open("/data/input.json") as f:
-        config_data = json.loads(f.read())
+    config_data = get_refinery_input_as_json()
+
+    for refinery_node_uuid in config_data["node_info"]:
+        TILE_SETS[refinery_node_uuid] = {}
```
Comment (mccalluc) [inline on_startup.py]: Feels weird to assign to something that should be a constant?

File: on_startup.py [hunk 5]
```diff
@@ -1,37 +1,69 @@
...
+
+        for key in refinery_node["node_solr_info"].keys():
+            if HIGLASS_FILE_TYPE in key.lower():
+                TILE_SETS[refinery_node_uuid][HIGLASS_FILE_TYPE] = (
+                    refinery_node["node_solr_info"][key]
+                )
+            if HIGLASS_DATA_TYPE in key.lower():
+                TILE_SETS[refinery_node_uuid][HIGLASS_DATA_TYPE] = (
+                    refinery_node["node_solr_info"][key]
+                )
```
Comment (mccalluc) [inline on_startup.py]: Maybe a comment about what typical values for keys are? I mistrust the `in key.lower()`. What behavior do you want if they're both true?

File: on_startup.py [hunk 6]
```diff
@@ -1,37 +1,69 @@
...
-        config_data = json.loads(f.read())
+    config_data = get_refinery_input_as_json()
+
+    for refinery_node_uuid in config_data["node_info"]:
+        TILE_SETS[refinery_node_uuid] = {}
+
+        refinery_node = config_data["node_info"][refinery_node_uuid]
+        TILE_SETS[refinery_node_uuid]["file_url"] = refinery_node["file_url"]
+        TILE_SETS[refinery_node_uuid]["file_name"] = refinery_node[
+            "file_url"].split("/")[-1]
```
Comment (mccalluc) [inline on_startup.py]: Might be more clear to make a dict and then make one assignment to `TILE_SETS[refinery_node_uuid]`

File: on_startup.py [hunk 7]
```diff
@@ -1,37 +1,69 @@
...
                 )
             )
         else:
-            with open('{}{}'.format(data_dir, url.split("/")[-1]), 'wb') as f:
+            with open(
+                '{}{}'.format(
+                    data_dir,
+                    TILE_SETS[refinery_node_uuid]["file_name"]
+                ), 'wb'
+            ) as f:
```
Comment (mccalluc) [inline on_startup.py]: Here and above, I think it was more clear having a short variable instead of `TILE_SETS[refinery_node_uuid]["file_name"]` and `TILE_SETS[refinery_node_uuid]["file_url"]`

File: on_startup.py [hunk 8]
```diff
@@ -42,69 +74,29 @@ def populate_higlass_data_directory(data_dir):
...
-    )
-
-    for filename in files_to_ingest:
+    Ingest previously downloaded files into higlass-server w/ django
+    management command
+    :param data_dir: <String> Path to directory populated with data to ingest
+    """
+    for refinery_node_uuid in TILE_SETS.keys():
+        refinery_node_info = TILE_SETS[refinery_node_uuid]
+        file_path = data_dir + refinery_node_info["file_name"]
```
Comment (mccalluc) [inline on_startup.py]: This comes goes into `filename`: Double check? changing kwarg would be out of line?

File: on_startup.py [hunk 9]
```diff
@@ -42,69 +74,29 @@ def populate_higlass_data_directory(data_dir):
...
     # Allows for django commands to run in a standalone script
     django.setup()
 
     populate_higlass_data_directory(data_dir)
     ingest_tilesets(data_dir)
 
-    # Don't switch page until data ingested
-    swap_waiting_page()
+    # Start Nginx only after tilesets have been ingested
+    subprocess.run(["/usr/sbin/nginx"])
```
Comment (mccalluc) [inline on_startup.py]: You sort of need to jump through hoops to completely hand off to a new process, but perhaps consider: https://stackoverflow.com/questions/3516007/run-process-and-dont-wait

(In perl `exec` did this, so I was wondering if there were an analog in python.)

File: test_runner.sh [hunk 10]
```diff
@@ -19,12 +19,27 @@ CONTAINER_NAME="container-$STAMP$SUFFIX"
...
+PYTHON_SERVER_PORT=9999
+
+get_local_url() {
+    local _ip _line
+    while IFS=$': \t' read -a _line ;do
+        [ -z "${_line%inet}" ] &&
+           _ip=${_line[${#_line[1]}>4?1:2]} &&
+           [ "${_ip#127.0.0.1}" ] && echo http://$_ip:$PYTHON_SERVER_PORT && return 0
+      done< <(LANG=C /sbin/ifconfig)
+}
```
Comment (mccalluc) [inline test_runner.sh:36]: Walk me through this.

File: test_runner.sh [hunk 11]
```diff
@@ -19,12 +19,27 @@ CONTAINER_NAME="container-$STAMP$SUFFIX"
...
+
+python -m SimpleHTTPServer $PYTHON_SERVER_PORT &
+PYTHON_SERVER_PID=$!
+docker run --env INPUT_JSON_URL=$(get_local_url)/test-data/input.json \
```
Comment (mccalluc) [inline test_runner.sh]: This is all just to get a static url? Maybe just reference a raw github resource?

File: tests.py [hunk 12]
```diff
@@ -2,40 +2,51 @@
...
+        os.environ['PORT'] = subprocess.check_output(
+            command, shell=True).strip().decode('utf-8')
+        self.base_url = "http://localhost:{PORT}/".format(**os.environ)
+        self.tilesets_url = '{}api/v1/tilesets/'.format(self.base_url)
         while True:
-            if 0 == subprocess.call('curl --fail --silent '+self.url+' > /dev/null', shell=True):
+            if 0 == subprocess.call(
+                'curl --fail --silent ' + self.tilesets_url + ' > /dev/null',
+                shell=True
+            ):
```
Comment (mccalluc) [inline tests.py]: I'm embarrassed to have written this code. See https://github.com/refinery-platform/refinery-developer-vis-tool/blob/master/test.py#L26-L42

File: tests.py [hunk 13]
```diff
@@ -2,40 +2,51 @@
...
                 break
             print('still waiting for server...')
-            time.sleep(1)
+            time.sleep(2)
 
-    def assertRun(self, command, res=[r'']):
-        output = subprocess.check_output(command.format(**os.environ), shell=True).strip()
+    def assert_run(self, command, res=[r'']):
```
Comment (mccalluc) [inline tests.py]: Does this actually get used in a real test?

**Review by mccalluc — COMMENTED** [review]
I don't see big conceptual problems, but I think there are a lot of little things to clean up.

File: Dockerfile (hunk 1 above)
Comment (scottx611x) [reply Dockerfile]: Oh, this is a hack to only allow the HiGlass `/app` view i.e. the fullscreen mode. Just overwriting the main `index.html`, static assets are already in the proper places at this point

File: Dockerfile [hunk 14]
```diff
@@ -26,12 +18,21 @@ RUN sed -i 's@"\/api\/v1",@"\.\/api\/v1"@g' \
...
 # Higlass currently has no favicon.png causing a 500 Error
 RUN touch higlass-website/assets/images/favicon.png
 
 ENV DJANGO_SETTINGS_MODULE="higlass_server.settings"
+
+COPY on_startup.py /home/higlass/projects/higlass-server
+
+# Append to the supervisord.conf and set the priority of `on_startup.py` to
+# be greater than the default of `999` so that it starts up last
+RUN ( echo ""; \
```
Comment (scottx611x) [inline Dockerfile]: just `echo;` ?

File: on_startup.py (hunk 9 above)
Comment (scottx611x) [reply on_startup.py]: Comment that parent will hang around

File: test_runner.sh (hunk 10 above)
Comment (scottx611x) [reply test_runner.sh:36]: Explain this in plain english and point to SO

File: tests.py [hunk 15]
```diff
@@ -7,35 +7,48 @@
...
+            "perl -pne 's/.*://'".format(**os.environ)
+        os.environ['PORT'] = subprocess.check_output(
+            command, shell=True).strip().decode('utf-8')
+        self.base_url = "http://localhost:{PORT}/".format(**os.environ)
+        self.tilesets_url = '{}api/v1/tilesets/'.format(self.base_url)
+
+        server_check_count = 0
+        while server_check_count <= 5:
+            if 0 == subprocess.call(
+                'curl --fail --silent ' + self.tilesets_url + ' > /dev/null',
```
Comment (scottx611x) [inline tests.py]: requests.get()

File: tests.py [hunk 16]
```diff
@@ -7,35 +7,48 @@
...
+                shell=True
+            ):
                 break
             print('still waiting for server...')
-            time.sleep(1)
+            time.sleep(5)
```
Comment (scottx611x) [inline tests.py]: 5 seems long

File: tests.py [hunk 17]
```diff
@@ -7,35 +7,48 @@
...
             self.assertRegexpMatches(output, re)
 
     # Tests:
     def test_hello(self):
-        self.assertRun('echo "hello?"', [r'hello'])
+        self.assert_run('echo "hello?"', [r'hello'])
+
+    def test_home_page(self):
+        response = requests.get(self.base_url)
+        self.assertEqual(response.status_code, 200)
```
Comment (scottx611x) [inline tests.py]: stronger assertions about the homepage

File: tests.py [hunk 18]
```diff
@@ -7,35 +7,48 @@
...
 
-    # Test if the data we specify in input.json gets ingested properly by
+    # Test if the data we specify in INPUT_JSON_URL gets ingested properly by
     # higlass server upon container startup
     def test_data_ingested(self):
         time.sleep(5)
```
Comment (scottx611x) [inline tests.py]: check this necessity

File: tests.py [hunk 19]
```diff
@@ -52,4 +65,3 @@ def test_data_ingested(self):
         print('PASS!')
     else:
         print('FAIL!')
-        exit(1)
```
Comment (scottx611x) [inline tests.py]: Definitely need this, maybe its worth moving simplehttpserver start & cleanup into python
